tensorflow-framework/src/test/java/org/tensorflow/framework/metrics/AccuracyTest.java

```diff
@@ -0,0 +1,130 @@
+      Operand<TInt32> labels = tf.reshape(tf.constant(trueArray), tf.constant(Shape.of(4, 1)));
+      Operand<TFloat32> predictions =
+          tf.reshape(tf.constant(predArray), tf.constant(Shape.of(4, 1)));
+      Op op = instance.updateState(labels, predictions, null);
+      session.run(op);
+      Variable<TFloat32> total = instance.getTotal();
+      Variable<TFloat32> count = instance.getCount();
+      Operand<TFloat32> result = instance.result();
+      session.evaluate(4F, total);
+      session.evaluate(4, count);
+      session.evaluate(1F, result);
+    }
+  }
+
+  @Test
+  public void testSampleWeight() {
+    try (TestSession session = TestSession.createTestSession(tfMode)) {
+      Ops tf = session.getTF();
+      Accuracy<TFloat32> instance = new Accuracy<>(tf, 1001L, TFloat32.class);
+      session.run(instance.resetStates());
+      float[] trueArray = {2, 1};
+      float[] predArray = {2, 0};
+
+      Operand<TFloat32> labels = tf.reshape(tf.constant(trueArray), tf.constant(Shape.of(2, 1)));
+      Operand<TFloat32> predictions =
+          tf.reshape(tf.constant(predArray), tf.constant(Shape.of(2, 1)));
+
+      Operand<TFloat32> sampleWeight =
+          tf.reshape(tf.constant(new float[] {.5F, .2F}), tf.constant(Shape.of(2, 1)));
+      Op op = instance.updateState(labels, predictions, sampleWeight);
+      session.run(op);
+
+      Variable<TFloat32> total = instance.getTotal();
+      Variable<TFloat32> count = instance.getCount();
+      Operand<TFloat32> result = instance.result();
+      session.evaluate(.5F, total);
+      session.evaluate(.7, count);
+      session.evaluate(0.71428573f, result);
+    }
+  }
+
+  @Test
+  public void testVariableState() {
+    try (TestSession session = TestSession.createTestSession(tfMode)) {
+      Ops tf = session.getTF();
+      Accuracy<TFloat32> instance = new Accuracy<>(tf, 1001L, TFloat32.class);
+      session.run(instance.resetStates());
+      float[] trueArray = {2, 1};
+
+      Operand<TFloat32> labels = tf.reshape(tf.constant(trueArray), tf.constant(Shape.of(2, 1)));
+
+      Operand<TFloat32> sampleWeight =
+          tf.reshape(tf.constant(new float[] {.5F, .2F}), tf.constant(Shape.of(2, 1)));
+      Op op = instance.updateState(labels, labels, sampleWeight);
+      session.run(op);
+      Variable<TFloat32> total = instance.getTotal();
+      Variable<TFloat32> count = instance.getCount();
+      Operand<TFloat32> result = instance.result();
+      session.evaluate(0.7F, total);
+      session.evaluate(.7, count);
+      session.evaluate(1.0F, result);
+
+      // 2nd run
+      op = instance.updateState(labels, labels, sampleWeight);
+      session.run(op);
+      result = instance.result();
+      session.evaluate(1.4F, total);
+      session.evaluate(1.4, count);
+      session.evaluate(1.0F, result);
+
+      // new instance same graph
+      instance = new Accuracy<>(tf, 1001L, TFloat32.class);
+      session.run(instance.resetStates());
```

Comment:
What are we trying to test with this sequence of four calls to `Session.run`? Are we confident that the semantics of `Session.run` will precisely emulate some particular production usage? Or can we write this code more like production code would be written?

In particular, I'm interested in the effect of our second combination of `Session.run` with `Accuracy.resetStates` below:
~~~Java
        // reset variables
        session.run(instance.resetStates());
~~~
Does that have exactly the same semantics as forming a control dependency on the `Op` returned by `resetStates`, or does it force a fresh execution of that `Op` even if it was previously executed?

Reply:
A metric with the same name will pick up the variables created by a previous metric with that same name.
It is intentional.
This test tests that this functionality works.

For example, in Reduce

total = getTF().withName(totalName).variable(Shape.scalar(), resultType);

Returns the variable with the same name, `totalName`, from the Graph it it already exists. Otherwise , it creates a new `Variable`.